src/main/java/org/checkerframework/checker/dependencyinjection/DependencyInjectionAnnotatedTypeFactory.java

```diff
@@ -256,11 +435,97 @@ public Void visitMethod(MethodTree tree, AnnotatedTypeMirror p) {
       ExecutableElement element = TreeUtils.elementFromDeclaration(tree);
       if (ElementUtils.hasAnnotation(element, Provides.class.getName())) {
         // TODO: Put fully qualified names of classes in knownBindings
-        knownBindings.put(tree.getReturnType().toString(), p.getUnderlyingType().toString());
+        knownBindings.put(
+            tree.getReturnType().toString(),
+            new KnownBindingsValue.KnownBindingsValueBuilder(p.getUnderlyingType().toString())
+                .build());
       }
 
       printKnownBindings();
       return super.visitMethod(tree, p);
     }
   }
+
+  @Override
+  protected QualifierHierarchy createQualifierHierarchy() {
+    return new DependencyInjectionQualifierHierarchy(
+        this.getSupportedTypeQualifiers(), this.elements);
+  }
+
+  protected class DependencyInjectionQualifierHierarchy extends AccumulationQualifierHierarchy {
+
+    private final AnnotationMirror bawAnno =
+        AnnotationBuilder.fromClass(elements, BindAnnotatedWith.class);
+
+    protected DependencyInjectionQualifierHierarchy(
+        Collection<Class<? extends Annotation>> qualifierClasses, Elements elements) {
+      super(qualifierClasses, elements);
+    }
+
+    @Override
+    public boolean isSubtype(final AnnotationMirror subAnno, final AnnotationMirror superAnno) {
+
+      if (AnnotationUtils.areSameByName(subAnno, bawAnno)
+          && AnnotationUtils.areSameByName(superAnno, bawAnno)) {
```

Comment:
Rather than using `areSameByName` on an `AnnotationMirror`, it is more efficient to just use `areSameByName(subAnno, "fully.qualifier.name.of.your.Annotation")`, because the two `AnnotationMirror` version also just eventually does string comparison (with several additional operations in between). These functions are the most expensive part of typechecking (they are invoked many, many times!), so keeping them fast is important for performance.

(What I usually do is put the name of the annotation in a `final static String` field of the qual hierarchy, and then use that everywhere.)

Reply:
Done.